File: polyglot.py
#!/usr/bin/env python3
"""
polyglot.py

Usage:
    python polyglot.py <cppFile>.cpp <pyFile>.py <outFile>.cpp

This script will:
 - verify the C++ file compiles (syntax check),
 - verify the Python file passes pyflakes (syntax/lint check),
 - produce a merged polyglot file containing both C++ and Python code.

It accepts the .cpp and .py arguments in either order.
"""

import sys
import subprocess
from pathlib import Path
import shutil

USAGE = "Usage: polyglot.py <cppFile>.cpp <pyFile>.py <outFile>.cpp\n"


def run_cmd(cmd, capture_output=True, text=True):
    """Run command list `cmd` and return CompletedProcess."""
    try:
        return subprocess.run(cmd, capture_output=capture_output, text=text, check=False)
    except FileNotFoundError as e:
        # command not found
        class Dummy:
            returncode = 127
            stdout = ""
            stderr = str(e)
        return Dummy()


def check_cpp_syntax(cpp_path: Path) -> bool:
    """
    Check C++ syntax using g++ with -fsyntax-only.
    Returns True if no syntax errors, False otherwise.
    """
    print("Trying to compile C++ program (syntax-only check)...")
    # -fsyntax-only checks syntax without producing an object/binary
    proc = run_cmd(["g++", "-fsyntax-only", str(cpp_path)])
    # Many compilers print errors to stderr
    output = (proc.stdout or "") + (proc.stderr or "")
    if proc.returncode != 0:
        print("Syntax Errors present in C++ file. Compiler output:")
        print(output)
        return False
    print("No syntax errors, C++ syntax check passed.")
    return True


def check_py_syntax(py_path: Path) -> bool:
    """
    Check Python syntax using pyflakes. If pyflakes isn't installed,
    prompt user to install it via pip.
    Returns True if no syntax errors (pyflakes exit code 0), False otherwise.
    """
    print("Checking Python file syntax with pyflakes...")
    # Try invoking pyflakes as a module first (works if installed for current python)
    proc = run_cmd([sys.executable, "-m", "pyflakes", str(py_path)])
    stdout = proc.stdout or ""
    stderr = proc.stderr or ""

    # Detect module-not-found style message
    module_missing = "No module named pyflakes" in stderr or ("pyflakes" in stderr and "not found" in stderr.lower()) or proc.returncode == 127

    if module_missing:
        print("pyflakes is not installed for this Python interpreter.")
        # Ask user if they want to install
        proceed = input("PyFlakes not installed. Proceed to install? [Y/n] ").strip()
        if proceed == "" or proceed.lower() == "y":
            print("Installing pyflakes using pip...")
            pip_proc = run_cmd([sys.executable, "-m", "pip", "install", "pyflakes"])
            pip_out = (pip_proc.stdout or "") + (pip_proc.stderr or "")
            if pip_proc.returncode != 0:
                print("Fatal: pip install failed. Install pip and/or pyflakes manually and re-run the script.")
                print("pip output:")
                print(pip_out)
                return False
            # re-run pyflakes
            proc = run_cmd([sys.executable, "-m", "pyflakes", str(py_path)])
            stdout = proc.stdout or ""
            stderr = proc.stderr or ""
        else:
            print("Aborting: pyflakes required to continue.")
            return False

    # If pyflakes returns non-zero, it might have found issues and printed them to stdout/stderr.
    if proc.returncode != 0:
        combined = (stdout or "") + (stderr or "")
        # pyflakes prints messages about issues; treat any output as "errors/warnings"
        if combined.strip():
            print("pyflakes reported issues:")
            print(combined)
            # treat as fatal as original code treats 'invalid syntax' as fatal
            return False
        else:
            # No output but nonzero? treat as failure.
            print("pyflakes returned non-zero exit code; output:")
            print(combined)
            return False

    # pyflakes exitcode == 0 => no issues
    print("No syntax errors, python syntax check passed.")
    return True


def merge_files(cpp_path: Path, py_path: Path, out_path: Path) -> None:
    """
    Merge the content using the polyglot wrapper:
    #if 0
    '''
    #endif
    <C++ code>
    #if 0
    '''
    #endif
    #if 0
    <python code>
    #endif
    """
    print("Merging files...")
    try:
        cpp_text = cpp_path.read_text(encoding="utf-8")
    except Exception as e:
        print(f"Failed to read C++ file: {e}")
        raise

    try:
        py_text = py_path.read_text(encoding="utf-8")
    except Exception as e:
        print(f"Failed to read Python file: {e}")
        raise

    merged = []
    merged.append("#if 0")
    merged.append("'''")
    merged.append("#endif")
    merged.append("")  # blank line
    merged.append(cpp_text.rstrip("\n"))  # preserve original except trailing newline
    merged.append("")  # ensure separation
    merged.append("#if 0")
    merged.append("'''")
    merged.append("#endif")
    merged.append("")  # blank line
    merged.append("#if 0")
    merged.append(py_text.rstrip("\n"))
    merged.append("#endif")

    out_path.write_text("\n".join(merged) + "\n", encoding="utf-8")
    print(f"Finished. Check {out_path}")


def main(argv):
    if len(argv) != 4:
        print("Error: expected 3 arguments, got", len(argv) - 1)
        print(USAGE, end="")
        return 1

    # argv[1], argv[2] may be in either order
    p1 = Path(argv[1])
    p2 = Path(argv[2])
    outp = Path(argv[3])

    if not p1.exists() or not p2.exists():
        print("Error: one or more input files do not exist.")
        print(f"Got: {p1} (exists={p1.exists()}), {p2} (exists={p2.exists()})")
        return 1

    # determine which is which
    if p1.suffix == ".cpp" and p2.suffix == ".py":
        cpp_path, py_path = p1, p2
    elif p1.suffix == ".py" and p2.suffix == ".cpp":
        py_path, cpp_path = p1, p2
    else:
        print("Usage not recognized.")
        print(USAGE, end="")
        return 1

    # C++ syntax check
    if not check_cpp_syntax(cpp_path):
        return 1

    # Python syntax check via pyflakes
    if not check_py_syntax(py_path):
        return 1

    # Merge
    try:
        merge_files(cpp_path, py_path, outp)
    except Exception as e:
        print("Failed to merge files:", e)
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: polyglot.cpp
#if 0
r'''
#endif

//main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <array>
#include <memory>
#include <filesystem>
#include <stdexcept>

namespace fs = std::filesystem;

std::string usageStr =
    "Usage: polyglot <source1> <source2> -o <outputFile>\n"
    "Supported extensions:\n"
    "  C/C++: .cpp, .cc, .cxx, .c\n"
    "  Python: .py\n"
    "  Ruby: .rb\n"
    "  Bash: .sh\n"
    "  Perl: .pl\n";  // 添加Perl支持说明
    
std::string runCmd(const std::string& cmd) {
    std::array<char, 256> buffer;
    std::string result;
    std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd.c_str(), "r"), pclose);
    if (!pipe) throw std::runtime_error("popen() failed!");
    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr)
        result += buffer.data();
    return result;
}

static std::string shellSafePath(const std::string& path) {
    std::string s = fs::path(path).generic_string();
    std::string escaped;
    for (char c : s) {
        if (c == '"') escaped += "\\\"";
        else escaped += c;
    }
    return "\"" + escaped + "\"";
}

std::string replace(const std::string& str, const std::string& replace, const std::string& with) {
    if (replace.empty()) return str;
    std::string result;
    result.reserve(str.size());
    std::size_t start = 0;
    std::size_t pos;
    while ((pos = str.find(replace, start)) != std::string::npos) {
        result.append(str, start, pos - start);
        result += with;
        start = pos + replace.length();
    }
    result.append(str, start, str.size() - start);
    return result;
}

bool checkSyntax(const std::string& file, const std::string& ext) {
    std::string res;
    std::string quoted = shellSafePath(file);

    if (ext == ".cpp" || ext == ".cc" || ext == ".cxx" || ext == ".c") {
        std::string flag = ext == ".c" ? "-x c " : "";
        res = runCmd("g++ -fsyntax-only " + flag + quoted + " 2>&1");
        if (!res.empty()) {
            std::cerr << "C/C++ syntax errors in " << file << ":\n" << res;
            return false;
        }
        return true;
    } else if (ext == ".py") {
        res = runCmd("pyflakes " + quoted + " 2>&1");
        if (!res.empty()) {
            std::string fallback = runCmd("python -m py_compile " + quoted + " 2>&1");
            if (!fallback.empty()) {
                std::cerr << "Python syntax errors in " << file << ":\n" << fallback;
                std::cerr << "If pyflakes is desired, please install it or ensure it's on PATH.\n";
                return false;
            }
        }
        return true;
    } else if (ext == ".rb") {
        res = runCmd("ruby -c " + quoted + " 2>&1");
        if (res.find("Syntax OK") == std::string::npos) {
            std::cerr << "Ruby syntax errors in " << file << ":\n" << res;
            return false;
        }
        return true;
    } else if (ext == ".sh") {
        res = runCmd("bash -n " + quoted + " 2>&1");
        if (!res.empty()) {
            std::cerr << "Bash syntax errors in " << file << ":\n" << res;
            return false;
        }
        return true;
    } else if (ext == ".pl") {  // 添加Perl语法检查
        res = runCmd("perl -c " + quoted + " 2>&1");
        if (res.find("syntax OK") == std::string::npos) {
            std::cerr << "Perl syntax errors in " << file << ":\n" << res;
            return false;
        }
        return true;
    }

    std::cerr << "\nUnsupported file extension: " << ext << "\n";
    return false;
}

std::vector<std::string> readFile(const std::string& filename) {
    std::ifstream in(filename);
    if (!in.is_open()) throw std::runtime_error("Failed to open: " + filename);
    std::vector<std::string> lines;
    std::string line;
    while (std::getline(in, line)) lines.push_back(line);
    return lines;
}

void writeMerged(
    const std::string& outFile,
    const std::string& ext1, const std::vector<std::string>& content1,
    const std::string& ext2, const std::vector<std::string>& content2
) {
    std::ofstream out(outFile, std::ios::binary);
    if (!out.is_open()) throw std::runtime_error("Failed to open output: " + outFile);

    auto openFence = [](const std::string& ext) -> std::string {
        if (ext == ".py") return "r'''";
        if (ext == ".rb") return "=begin";
        if (ext == ".sh") return ": '";
        if (ext == ".pl") return "=pod";  // 添加Perl围栏开始
        return "";
    };

    auto closeFence = [](const std::string& ext) -> std::string {
        if (ext == ".py") return "'''";
        if (ext == ".rb") return "=end";
        if (ext == ".sh") return "'";
        if (ext == ".pl") return "=cut";  // 添加Perl围栏结束
        return "";
    };

    auto escapeCpp = [](const std::string& content) -> std::string {
        return "#if 0\n" + content + "\n#endif\n";
    };

    auto writeLine = [&out](const std::string& line) {
        out.write(line.c_str(), line.size());
        out.put('\n');
    };

    auto escapeForPython = [](const std::string& line) -> std::string {
        std::string out;
        for (size_t i = 0; i < line.size(); i++) {
            if (i + 2 < line.size() && line[i] == '\'' && line[i+1] == '\'' && line[i+2] == '\'') {
                out += "\\'\\'\\'";
                i += 2;
            } else {
                out += line[i];
            }
        }
        return out;
    };

    if (ext1 == ".cpp" || ext1 == ".cc" || ext1 == ".cxx" || ext1 == ".c") {
        writeLine(escapeCpp(openFence(ext2)));
        for (const std::string& l : content1) {
            writeLine(escapeForPython(l));
        }
        writeLine(escapeCpp(closeFence(ext2)));

        writeLine("#if 0");
        for (auto& l : content2) {
            writeLine(escapeForPython(l));
        }
        writeLine("#endif");
    }
    else if (ext2 == ".cpp" || ext2 == ".cc" || ext2 == ".cxx" || ext2 == ".c") {
        writeLine(escapeCpp(openFence(ext1)));
        for (auto& l : content2) writeLine(l);
        writeLine(escapeCpp(closeFence(ext1)));

        writeLine("#if 0");
        for (auto& l : content1) writeLine(l);
        writeLine("#endif");
    }
    else {
        throw std::runtime_error("No C/C++ file in pair");
    }
}

int main(int argc, char* argv[]) {
    std::vector<std::string> args(argv, argv + argc);
    if (argc < 5) {
        std::cerr << usageStr;
        return 1;
    }

    std::string file1, file2, outFile;
    for (int i = 1; i < argc; i++) {
        if (args[i] == "-o") {
            if (i + 1 >= argc) {
                std::cerr << "Error: -o requires an argument\n";
                return 1;
            }
            outFile = args[++i];
        } else if (file1.empty()) {
            file1 = args[i];
        } else if (file2.empty()) {
            file2 = args[i];
        } else {
            std::cerr << "Error: unexpected argument: " << args[i] << "\n";
            return 1;
        }
    }

    if (file1.empty() || file2.empty() || outFile.empty()) {
        std::cerr << usageStr;
        return 1;
    }

    std::string ext1 = fs::path(file1).extension().string();
    std::string ext2 = fs::path(file2).extension().string();

    std::cout << "Checking syntax for " << file1 << "... ";
    if (!checkSyntax(file1, ext1)) {
        std::cerr << "\nSyntax error in " << file1 << "\n";
        return 1;
    }
    std::cout << "OK\n";

    std::cout << "Checking syntax for " << file2 << "... ";
    if (!checkSyntax(file2, ext2)) {
        std::cerr << "\nSyntax error in " << file2 << "\n";
        return 1;
    }
    std::cout << "OK\n";

    auto content1 = readFile(file1);
    auto content2 = readFile(file2);

    writeMerged(outFile, ext1, content1, ext2, content2);
    std::cout << "Merged into " << outFile << "\n";
    return 0;
}

#if 0
'''
#endif

#if 0
import subprocess
import sys
import os
from pathlib import Path

usage_str = """Usage: polyglot <source1> <source2> -o <outputFile>
Supported extensions:
  C/C++: .cpp, .cc, .cxx, .c
  Python: .py
  Ruby: .rb
  Bash: .sh
  Perl: .pl"""

def run_cmd(cmd):
    try:
        result = subprocess.run(cmd, shell=True, check=True, 
                              stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                              text=True)
        return result.stdout
    except subprocess.CalledProcessError as e:
        return e.stderr

def shell_safe_path(path):
    path_str = str(Path(path))
    escaped = path_str.replace('"', '\\"')
    return f'"{escaped}"'

def check_syntax(file, ext):
    quoted = shell_safe_path(file)
    
    if ext in ('.cpp', '.cc', '.cxx', '.c'):
        flag = '-x c ' if ext == '.c' else ''
        res = run_cmd(f"g++ -fsyntax-only {flag}{quoted}")
        if res.strip():
            print(f"C/C++ syntax errors in {file}:\n{res}", file=sys.stderr)
            return False
        return True
    elif ext == '.py':
        res = run_cmd(f"pyflakes {quoted}")
        if res.strip():
            fallback = run_cmd(f"python -m py_compile {quoted}")
            if fallback.strip():
                print(f"Python syntax errors in {file}:\n{fallback}", file=sys.stderr)
                print("If pyflakes is desired, please install it or ensure it's on PATH.", file=sys.stderr)
                return False
        return True
    elif ext == '.rb':
        res = run_cmd(f"ruby -c {quoted}")
        if "Syntax OK" not in res:
            print(f"Ruby syntax errors in {file}:\n{res}", file=sys.stderr)
            return False
        return True
    elif ext == '.sh':
        res = run_cmd(f"bash -n {quoted}")
        if res.strip():
            print(f"Bash syntax errors in {file}:\n{res}", file=sys.stderr)
            return False
        return True
    elif ext == '.pl':
        res = run_cmd(f"perl -c {quoted}")
        if "syntax OK" not in res:
            print(f"Perl syntax errors in {file}:\n{res}", file=sys.stderr)
            return False
        return True
    
    print(f"\nUnsupported file extension: {ext}\n", file=sys.stderr)
    return False

def read_file(filename):
    with open(filename, 'r', encoding='utf-8') as f:
        return f.readlines()

def write_merged(out_file, ext1, content1, ext2, content2):
    def open_fence(ext):
        if ext == '.py': return "r'''"
        if ext == '.rb': return "=begin"
        if ext == '.sh': return ": '"
        if ext == '.pl': return "=pod"
        return ""
    
    def close_fence(ext):
        if ext == '.py': return "'''"
        if ext == '.rb': return "=end"
        if ext == '.sh': return "'"
        if ext == '.pl': return "=cut"
        return ""
    
    def escape_cpp(content):
        return "#if 0\n" + content + "\n#endif\n"
    
    def escape_for_python(line):
        return line.replace("'''", "\\'\\'\\'")
    
    with open(out_file, 'w', encoding='utf-8') as out:
        def write_line(line):
            out.write(line + '\n')
        
        if ext1 in ('.cpp', '.cc', '.cxx', '.c'):
            write_line(escape_cpp(open_fence(ext2)))
            for line in content1:
                write_line(escape_for_python(line.rstrip('\n')))
            write_line(escape_cpp(close_fence(ext2)))
            
            write_line("#if 0")
            for line in content2:
                write_line(escape_for_python(line.rstrip('\n')))
            write_line("#endif")
        elif ext2 in ('.cpp', '.cc', '.cxx', '.c'):
            write_line(escape_cpp(open_fence(ext1)))
            for line in content2:
                write_line(line.rstrip('\n'))
            write_line(escape_cpp(close_fence(ext1)))
            
            write_line("#if 0")
            for line in content1:
                write_line(line.rstrip('\n'))
            write_line("#endif")
        else:
            raise RuntimeError("No C/C++ file in pair")

def main():
    args = sys.argv[1:]
    if len(args) < 4:
        print(usage_str, file=sys.stderr)
        sys.exit(1)
    
    file1 = file2 = out_file = None
    i = 0
    while i < len(args):
        if args[i] == '-o':
            if i + 1 >= len(args):
                print("Error: -o requires an argument", file=sys.stderr)
                sys.exit(1)
            out_file = args[i+1]
            i += 2
        elif file1 is None:
            file1 = args[i]
            i += 1
        elif file2 is None:
            file2 = args[i]
            i += 1
        else:
            print(f"Error: unexpected argument: {args[i]}", file=sys.stderr)
            sys.exit(1)
    
    if not file1 or not file2 or not out_file:
        print(usage_str, file=sys.stderr)
        sys.exit(1)
    
    ext1 = Path(file1).suffix
    ext2 = Path(file2).suffix
    
    print(f"Checking syntax for {file1}... ", end='')
    if not check_syntax(file1, ext1):
        print(f"\nSyntax error in {file1}", file=sys.stderr)
        sys.exit(1)
    print("OK")
    
    print(f"Checking syntax for {file2}... ", end='')
    if not check_syntax(file2, ext2):
        print(f"\nSyntax error in {file2}", file=sys.stderr)
        sys.exit(1)
    print("OK")
    
    content1 = read_file(file1)
    content2 = read_file(file2)
    
    write_merged(out_file, ext1, content1, ext2, content2)
    print(f"Merged into {out_file}")

if __name__ == "__main__":
    main()
#endif
